supervisor/options.py

```diff
@@ -1395,7 +1397,6 @@ def drop_privileges(self, user):
             os.setgid(gid)
         except OSError:
             return 'Could not set group id of effective user'
-        os.setuid(uid)
```

Comment:
Why specifically are you trying to move `os.setuid()` before setting `os.setgroups()` and `os.setgid()`?

Reply:
with previous code this the place where the code used to break if there is a problem with setuid. So, I moved setuid before so that if there is a problem with setuid it breaks at the same place. In fact, the order setuid or setgid should not matter.

Reply:
Sorry, I don't understand this explanation.  If I am reading correctly, you are saying that you moved it because it was broken in the place that it was, but then you also said you think the order does not matter.

Reply:
https://github.com/Supervisor/supervisor/pull/1402#issuecomment-772720968:
> 3) moved setuid before because to detect failure at earliest please. In other words of kept at end means, function performs other tasks only to fail at end, not a good idea!

I don't understand how it could work if `os.setuid()` is called before `os.setgroups()` and `os.setgid()` as you propose here.  I think the order is important and that the order needs to be as it was originally.

Before any code changes:

```text
$ id unprivileged 
uid=1001(unprivileged) gid=1001(unprivileged) groups=1001(unprivileged)

$ id 
uid=0(root) gid=0(root) groups=0(root)

$ whoami 
root 

$ cat supervisord.conf 
[supervisord]
nodaemon=true 
loglevel=debug 

[program:id]
user = unprivileged
command = /usr/bin/id 
startsecs = 0 

$ supervisord -c supervisord.conf 
2021-02-03 11:02:35,141 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2021-02-03 11:02:35,145 INFO supervisord started with pid 2663
2021-02-03 11:02:36,148 INFO spawned: 'id' with pid 2666
2021-02-03 11:02:36,158 DEBG 'id' stdout output:
uid=1001(unprivileged) gid=1001(unprivileged) groups=1001(unprivileged)
```

Making no other changes, other than to move `os.setuid()` before `os.setgroups()` and `os.setgid()` as you have done in this pull request:

```diff
diff --git a/supervisor/options.py b/supervisor/options.py
index 273d8a5..c83fe44 100644
--- a/supervisor/options.py
+++ b/supervisor/options.py
@@ -1375,6 +1375,8 @@ class ServerOptions(Options):
         if current_uid != 0:
             return "Can't drop privilege as nonroot user"
 
+        os.setuid(uid)
+
         gid = pwrec[3]
         if hasattr(os, 'setgroups'):
             user = pwrec[0]
@@ -1395,7 +1397,6 @@ class ServerOptions(Options):
             os.setgid(gid)
         except OSError:
             return 'Could not set group id of effective user'
-        os.setuid(uid)
 
     def set_uid_or_exit(self):
         """Set the uid of the supervisord process.  Called during supervisord
```

Results in the call to `os.setgroups()` failing:

```
$ whoami 
root 

$ supervisord -c supervisord.conf 
2021-02-03 11:05:07,287 CRIT Supervisor is running as root.  Privileges were not dropped because no user is specified in the config file.  If you intend to run as root, you can set user=root in the config file to avoid this message.
2021-02-03 11:05:07,290 INFO supervisord started with pid 2706
2021-02-03 11:05:08,293 INFO spawned: 'id' with pid 2709
2021-02-03 11:05:08,300 DEBG 'id' stderr output:
supervisor: couldn't setuid to 1001: Could not set groups of effective user

2021-02-03 11:05:08,301 INFO success: id entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
2021-02-03 11:05:08,301 DEBG 'id' stderr output:
supervisor: child process was not spawned

2021-02-03 11:05:08,303 DEBG fd 8 closed, stopped monitoring <POutputDispatcher at 140498999981840 for <Subprocess at 140499000370624 with name id in state RUNNING> (stderr)>
2021-02-03 11:05:08,303 DEBG fd 6 closed, stopped monitoring <POutputDispatcher at 140499007533408 for <Subprocess at 140499000370624 with name id in state RUNNING> (stdout)>
2021-02-03 11:05:09,306 INFO exited: id (exit status 127; not expected)
2021-02-03 11:05:09,306 DEBG received SIGCLD indicating a child quit
```

Reply:
yes, you got it right. Apologies I overlooked that previous check. the setuid must be called only after fetching all needed details of the current user. I put the setuid at the end as before.